include/pybind11/detail/class.h

```diff
@@ -388,15 +388,18 @@ inline void clear_patients(PyObject *self) {
     auto *instance = reinterpret_cast<detail::instance *>(self);
     auto &internals = get_internals();
     auto pos = internals.patients.find(self);
+
     assert(pos != internals.patients.end());
-    // Clearing the patients can cause more Python code to run, which
-    // can invalidate the iterator. Extract the vector of patients
-    // from the unordered_map first.
-    auto patients = std::move(pos->second);
-    internals.patients.erase(pos);
-    instance->has_patients = false;
-    for (PyObject *&patient : patients) {
-        Py_CLEAR(patient);
+    if (pos != internals.patients.end()) {
```

Comment:
This makes it a silent failure (if `NDEBUG` is not defined), which is actually worse than a segfault or similar.

I think it'll be best to throw an exception, i.e. simply replace the `assert` with something like:

```
    if (pos == internals.patients.end()) {
        pybind11_fail("FATAL: Internal consistency check failed: Invalid clear_patients() call.");
    }
```

Reply:
Yes, this is what I (tried to) suggest.

Reply:
Thanks, and sorry for the confusion.